### PR TITLE
View comments BE

### PR DESCRIPTION
**Pull request by dayanamdr**

### CODE REVIEW

File: codewellBe/lib/codewellBE.pm [hunk 1]
```diff
@@ -49,6 +49,8 @@ sub startup {
   # Comment
   $if_login->post('/api/comment/add')->name('on_comment_add')
     ->to('comment#on_add');
+  $if_login->get('/api/comment/view/:code_id')->name('on_comment_view')
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE.pm:52]: /api/comments/:code_id

File: codewellBe/lib/codewellBE/Controller/Comment.pm [hunk 2]
```diff
@@ -56,4 +57,20 @@ sub on_add {
   $self->render(json => { success => $success, error => $error });
 }
 
+sub view_comments {
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Controller/Comment.pm:60]: comments

File: codewellBe/lib/codewellBE/Controller/Comment.pm [hunk 3]
```diff
@@ -56,4 +57,20 @@ sub on_add {
...
+  my $self = shift;
+  my $code_id = $self->stash('code_id');
+
+  my ($success, $error) = (1, '');
+  my $code = validate_code_existence($self, $code_id);
+
+  $error .= 'Code id does not exist.' unless $code;
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Controller/Comment.pm:67]: we have a hash that contains all the errors, check Helper/Errors.pm

File: codewellBe/lib/codewellBE/Controller/Comment.pm [hunk 4]
```diff
@@ -56,4 +57,20 @@ sub on_add {
...
+  $success = 0 if $error ne '';
+
+  my $comments = get_comments($self, $code_id) if $success;
+
+  $self->render(json => { success => $success, error => $error,
+    comments => $comments });
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Controller/Comment.pm:73]:
```suggestion
  $self->render(json => { 
    success => $success, 
    error => $error,
    comments => $comments 
  });
```

File: codewellBe/lib/codewellBE/Helper/Comment.pm [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package codewellBE::Helper::Comment;
+
+use strict;
+use warnings;
+
+use Exporter 'import';
+our @EXPORT = qw/get_comments/;
+our $VERSION = 1;
+
+sub get_comments {
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Helper/Comment.pm:10]: get_code_parent_comments

File: codewellBe/lib/codewellBE/Helper/Comment.pm [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  my ($self, $code_id) = @_;
+  return [ map {
+    {
+      user_id => $_->user_id,
+      date_posted => $_->date_posted,
+      body => $_->body
+    }
+  } $self->db->resultset('Comment')->ordered_by('date_posted', '-asc')
+      ->by_code_id($code_id)->all
+  ];
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Helper/Comment.pm:20]: this don't need to be on a new line

File: codewellBe/lib/codewellBE/Schema/ResultSet/Comment.pm [hunk 7]
```diff
@@ -9,4 +9,17 @@ sub by_id {
   return shift->find({ id => pop });
 }
 
+sub by_code_id {
+  my ($self, $code_id) = @_;
+  return shift->search({ code_id => $code_id });
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Schema/ResultSet/Comment.pm:14]:
```suggestion
  return shift->search({ code_id => pop });
```

File: codewellBe/t/Controller/comment.t [hunk 8]
```diff
@@ -172,4 +172,42 @@ subtest 'Comment add' => sub {
   }
 };
 
+my $code_id = create_code($t, "Hello World!")->id;
```
Comment (andreisamoila74) [inline codewellBe/t/Controller/comment.t:175]: there is already a created code

File: codewellBe/t/Controller/comment.t [hunk 9]
```diff
@@ -172,4 +172,42 @@ subtest 'Comment add' => sub {
...
+
+  $t->get_ok("/api/comment/view/inexistent")->status_is(200)
+    ->json_is({
+      success => 0,
+      error => 'Code id does not exist.',
+      comments => undef
+    });
+};
+
+subtest 'Can view comments' => sub {
```
Comment (andreisamoila74) [inline codewellBe/t/Controller/comment.t:198]: all these 3 subtests can be in a single one, called 'Code comments'

**Review by andreisamoila74 — CHANGES_REQUESTED** [review]

File: codewellBe/lib/codewellBE/Controller/Comment.pm [hunk 10]
```diff
@@ -56,4 +57,23 @@ sub on_add {
...
 }
 
+sub comments {
+  my $self = shift;
+  my $code_id = $self->stash('code_id');
+
+  my ($success, $error) = (1, '');
+  my $code = validate_code_existence($self, $code_id);
+
+  $error .= get_error("invalid_code_id") unless $code;
```
Comment (S7012MY) [inline codewellBe/lib/codewellBE/Controller/Comment.pm:67]: use single ticks (') for constant strings

File: codewellBe/lib/codewellBE/Helper/Comment.pm [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+sub get_code_parent_comments {
+  my ($self, $code_id) = @_;
+  return [ map {
+    {
+      user_id => $_->user_id,
+      date_posted => $_->date_posted,
+      body => $_->body
+    }
+  } $self->db->resultset('Comment')->ordered_by('date_posted', '-asc')
+      ->by_code_id($code_id)->all ];
```
Comment (S7012MY) [inline codewellBe/lib/codewellBE/Helper/Comment.pm:19]: filter first by code_id and then order them

File: codewellBe/t/Controller/comment.t [hunk 12]
```diff
@@ -172,4 +172,42 @@ subtest 'Comment add' => sub {
...
+    { user_id =>  $comment->user_id, date_posted => $comment->date_posted,
+      body => $comment->body }
+  );
+
+  # Can view comments
+  $t->get_ok("/api/comments/$code_id")->status_is(200)
+    ->json_is({
+      success => 1,
+      error => '',
+      comments => \@ok_comment
```
Comment (S7012MY) [inline codewellBe/t/Controller/comment.t:209]: this shouldn't be an array ref but an json array containing the comments.
Also test with multiple comments so the sorting is also tested

**Review by S7012MY — CHANGES_REQUESTED** [review]

File: codewellBe/lib/codewellBE.pm [hunk 13]
```diff
@@ -49,6 +49,8 @@ sub startup {
   # Comment
   $if_login->post('/api/comment/add')->name('on_comment_add')
     ->to('comment#on_add');
+  $if_login->get('/api/comments/:code_id')->name('on_comment_view')
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE.pm:52]: on_comments_view

File: codewellBe/lib/codewellBE/Schema/ResultSet/Comment.pm [hunk 14]
```diff
@@ -9,4 +9,17 @@ sub by_id {
   return shift->find({ id => pop });
 }
 
+sub by_code_id {
+  my ($self, $code_id) = @_;
```
Comment (andreisamoila74) [inline codewellBe/lib/codewellBE/Schema/ResultSet/Comment.pm:13]: you don't need this line anymore, that's why you used shift and pop

File: codewellBe/t/Controller/comment.t [hunk 15]
```diff
@@ -172,4 +173,57 @@ subtest 'Comment add' => sub {
...
+  $t->get_ok("/api/comments/$code_id")->status_is(200)
+    ->json_is({
+      success => 0,
+      error => 'Not enough permissions. Please log in first!'
+    });
+
+  login_user($t, 'user');
+
+  # Can't view the comments for an inexistent code_id
+  $t->get_ok("/api/comments/inexistent")->status_is(200)
```
Comment (andreisamoila74) [inline codewellBe/t/Controller/comment.t:192]:
```suggestion
  $t->get_ok('/api/comments/inexistent')->status_is(200)
```

File: codewellBe/t/Controller/comment.t [hunk 16]
```diff
@@ -172,4 +173,57 @@ subtest 'Comment add' => sub {
...
+    ->json_is({
+      success => 0,
+      error => 'Comment code id does not exist. ',
+      comments => undef
+    });
+
+  # Check the comments for a logged in user and a valid code_id
+  my $comment = create_comment($t, "First comment", $code_id);
+  my $second_comment = create_comment($t, "Second comment", $code_id);
```
Comment (andreisamoila74) [inline codewellBe/t/Controller/comment.t:201]: '' instead of ""

File: codewellBe/t/Controller/comment.t [hunk 17]
```diff
@@ -172,4 +173,57 @@ subtest 'Comment add' => sub {
...
+
+  my @ok_comments = ({
+      user_id =>  $comment->user_id,
+      date_posted => $comment->date_posted,
+      body => $comment->body
+    }, {
+      user_id =>  $second_comment->user_id,
+      date_posted => $second_comment->date_posted,
+      body => $second_comment->body
+    });
```
Comment (andreisamoila74) [inline codewellBe/t/Controller/comment.t:219]:
```suggestion
  my @ok_comments = (
    {
      user_id =>  $comment->user_id,
      date_posted => $comment->date_posted,
      body => $comment->body
    }, 
    {
      user_id =>  $second_comment->user_id,
      date_posted => $second_comment->date_posted,
      body => $second_comment->body
    }
  );
```
```suggestion
  my @ok_comments = ({
      user_id =>  $comment->user_id,
      date_posted => $comment->date_posted,
      body => $comment->body
    }, {
      user_id =>  $second_comment->user_id,
      date_posted => $second_comment->date_posted,
      body => $second_comment->body
    });
```

**Review by andreisamoila74 — CHANGES_REQUESTED** [review]

**Review by andreisamoila74 — APPROVED** [review]

**Review by S7012MY — APPROVED** [review]

**Pull request merged**